### PR TITLE
Proof of Concept of container scanning

### PR DESCRIPTION
**Pull request by another-rex**
PoC of container scanning using stereoscope. The feature is currently behind the `--experimental-oci-image` flag.

### CODE REVIEW

File: internal/image/image.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	ctx := context.Background() // TODO: use proper context
+	img, err := stereoscope.GetImage(ctx, imagePath)
+	if err != nil {
+		return ScanResults{}, fmt.Errorf("failed to open image %s: %w", imagePath, err)
+	}
+
+	fileTree := img.SquashedTree()
+	allFiles := fileTree.AllFiles()
+
+	slices.SortFunc(allFiles, func(a, b file.Reference) int {
```
Comment (oliverchang) [inline internal/image/image.go:41]: add a comment explaining why we are sorting.

File: internal/image/image.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		return lockfile.Lockfile{}, fmt.Errorf("%w for %s", lockfile.ErrExtractorNotFound, f.Path())
+	}
+
+	packages, err := extractor.Extract(f)
+	log.Printf("%v\n", packages)
+	if err != nil && extractedAs != "" {
+		err = fmt.Errorf("(extracting as %s) %w", extractedAs, err)
+	}
+
+	sort.Slice(packages, func(i, j int) bool {
```
Comment (oliverchang) [inline internal/image/image.go:94]: Add a comment explaining why we sort.

File: pkg/osvscanner/osvscanner.go [hunk 3]
```diff
@@ -331,6 +333,33 @@ func (m *gitIgnoreMatcher) match(absPath string, isDir bool) (bool, error) {
 	return m.matcher.Match(pathInGitSep, isDir), nil
 }
 
+func scanImage(r reporter.Reporter, path string) ([]scannedPackage, error) {
```
Comment (oliverchang) [inline pkg/osvscanner/osvscanner.go:336]: What does the table output look like for this?

**Review by oliverchang — COMMENTED** [review]
nice!

File: docs/github-action.md [hunk 4]
```diff
@@ -161,6 +161,7 @@ The GitHub Actions have the following optional inputs:
   Default: `results.sarif`
 - `download-artifact`: Optional artifact to download for scanning. Can be used if you need to do some preprocessing to prepare the lockfiles for scanning. If the file names in the artifact are not standard lockfile names, make sure to add custom scan-args to specify the lockfile type and path (see [specify lockfiles](./usage#specify-lockfiles)).
 - `upload-sarif`: Whether to upload the results to Security > Code Scanning. Defaults to `true`.
+- `fail-on-vuln`: Whether to fail the workflow when a vulnerability is found. Defaults to `true`.
```
[the remark this answers is not in the document]
Comment (another-rex) [reply docs/github-action.md:164]: Yep, moved to https://github.com/google/osv-scanner/pull/810 , the changes are already published in docs, just haven't been merged into main yet.

File: pkg/lockfile/apk-installed.go [hunk 5]
```diff
@@ -109,3 +109,8 @@ func FromApkInstalled(pathToInstalled string) (Lockfile, error) {
 		Packages: packages,
 	}, err
 }
+
+//// nolint:gochecknoinits
```
[the remark this answers is not in the document]
Comment (another-rex) [reply pkg/lockfile/apk-installed.go:113]: apk-installed was never registered as an extractor (same with dpkg), there is a special branch during scanning to use the apk-installed extractor. (You need to manually specify `apk-installed:path/to/apk-installed`)

I'll remove this comment here, it's left in from me experimenting with registering the extractor.

File: pkg/osvscanner/osvscanner.go (hunk 3 above)
Comment (another-rex) [reply pkg/osvscanner/osvscanner.go:336]: Added a integration test, it looks like this:

```
+--------------------------------+------+-----------+---------+-----------+--------------------------------------------------------------+
| OSV URL                        | CVSS | ECOSYSTEM | PACKAGE | VERSION   | SOURCE                                                       |
+--------------------------------+------+-----------+---------+-----------+--------------------------------------------------------------+
| https://osv.dev/CVE-2018-25032 | 7.5  | Alpine    | zlib    | 1.2.11-r1 | ./fixtures/oci-image/alpine-tester.tar:/lib/apk/db/installed |
| https://osv.dev/CVE-2022-37434 | 9.8  | Alpine    | zlib    | 1.2.11-r1 | ./fixtures/oci-image/alpine-tester.tar:/lib/apk/db/installed |
+--------------------------------+------+-----------+---------+-----------+--------------------------------------------------------------+
```

**Review by oliverchang — APPROVED** [review]

**Review by hogo6002 — APPROVED** [review]

**Comment by oliverchang** [comment]
Rex and I discussed offline that we can likely leverage the lower level https://pkg.go.dev/github.com/google/go-containerregistry@v0.19.0/pkg/v1/tarball#ImageFromPath and do something similar to https://github.com/google/go-containerregistry/blob/8dadbe76ff8c20d0e509406f04b7eade43baa6c1/pkg/v1/mutate/mutate.go#L264 instead of using stereoscope. 

(can be for a future PR).

**Pull request merged**